psearch.1
.TH PSEARCH "1" "June 2012" "psearch 2.0.2"
.SH NAME
psearch \- search the FreeBSD ports
.SH SYNOPSIS
\fBpsearch\fR [\fIoptions\fR] \fIPATTERN\fR [\fIPATTERN\fR \fI...\fR]
.SH DESCRIPTION
Lists ports whose description matches \fIPATTERN\fR. \fIPATTERN\fR is a
case-insensitive regular expression.
.PP
If more than one pattern is given, a port description has to match all of them
for the port to be listed. If \fB\-o\fR is given, a single pattern match
suffices instead.
.PP
By default, the name and the short description are searched. If \fB\-s\fR is
given, the long description is searched as well, which slows things down a
lot. This can be compensated for by limiting searches to a port category with
\fB\-c\fR, thus reducing the number of pkg-plist files that need to be
searched.
.PP
\fBpsearch\fR uses IEEE Std 1003.2 (``POSIX.2'') regular expressions as explained by 
.BR re_format (7) .
.SH OPTIONS
.TP
\fB\-V\fR, \fB\-\-version\fR
Show program's version number and exit.
.TP
\fB\-h\fR, \fB\-\-help\fR
Show a brief help message explaining the options and exit.
.TP
\fB\-c\fR \fICATEGORY\fR, \fB\-\-category\fR=\fICATEGORY\fR
Only search for ports in \fICATEGORY\fR. Speeds up searching,
especially when \fB\-\-search_long\fR is also specified.
.TP
\fB\-f\fR \fIFILE\fR, \fB\-\-file\fR=\fIFILE\fR
Path to INDEX file. Defaults to the standard location of the INDEX file
on the FreeBSD system that \fBpsearch\fR runs on. Non-standard locations
given in /etc/make.conf are ignored.
.TP
\fB\-l\fR, \fB\-\-long\fR
Display the long description (pkg\-descr file) for any match found. They are
not searched however, unless \fB\-\-search_long\fR is also given.
.TP
\fB\-m\fR, \fB\-\-maintainer\fR
Display the maintainer's email address instead of the short description
for any match found, and switch on searching the maintainer's email address.
Specifying this parameter and searching for \fBports@freebsd.org\fR displays
all unmaintained ports.
.TP
\fB\-n\fR, \fB\-\-name\fR
Print canonical name of a port, including its version.
.TP
\fB\-o\fR, \fB\-\-or\fR
List ports whose description match any \fIPATTERN\fR, instead of requiring it
to match all of them.
.TP
\fB\-s\fR, \fB\-\-search_long\fR
Search long descriptions (pkg\-descr file), which slows down
searching. The long descriptions are not displayed however, unless
\fB\-\-long\fR is also given.
.TP
\fB\-v\fR \fIINVERSE_PATTERN\fR, \fB\-\-inverse\fR=\fIINVERSE_PATTERN\fR
Searches for ports that do not match \fIINVERSE_PATTERN\fR. May be
specified several times.
.SH SEE ALSO
.BR re_format (7)
.SH BUGS
Any relevant values stored in environment variables or /etc/make.conf are ignored.
This means that if your INDEX file does not reside in /usr/ports, you'll need to use
the \fB-f\fR option when running \fBpsearch\fR.
.SH AUTHOR
Benjamin Lutz


psearch.cpp
#include <cerrno>
#include <cstdlib>
#include <cstring>
#include <getopt.h>
#include <iostream>
#include <regex.h>
#include <set>
#include <string>
#include <vector>
extern "C" {
#include <sys/types.h>
#ifdef __FreeBSD__
#include <sys/sysctl.h>
#endif
}

#include "index.h"

using namespace std;

void print_copyright();
void print_help(char* progname, const string &index_filename);

bool add_pattern(vector<regex_t> &patterns, const char *pattern);

/** Compares sets a and b to see whether they are disjoint. */
bool disjoint(const set<string>& a, const set<string>& b);

bool regex_search(const string &s, const regex_t &r);
bool regex_search_file(const string &path, const regex_t &r);

/**
 * Finds the default index filename depending on the OS version
 * we're running on.
 */
void set_default_index_filename(string &index_filename);

const string PROGNAME("psearch");
const string VERSION("2.1.0");
const string COPYRIGHT("Copyright 2008, 2012, 2020");
const string AUTHOR("Benjamin Lutz (mail");
const string AUTHOR2("maxlor.com)");


int main(int argc, char** argv) {
	bool flag_long = false;
	bool flag_maintainer = false;
	bool flag_name = false;
	bool flag_or = false;
	bool flag_search_long = false;
	string index_filename;
	set<string> categories;
	vector<regex_t> patterns;
	vector<regex_t> inverse_patterns;
	
	set_default_index_filename(index_filename);

	if (argc == 1) {
		print_help(argv[0], index_filename);
		return 0;
	}
	
	const struct option longopts[] = {
		{ "version",     no_argument,       NULL, 'V' },
		{ "help",        no_argument,       NULL, 'h' },
		{ "category",    required_argument, NULL, 'c' },
		{ "file",        required_argument, NULL, 'f' },
		{ "long",        no_argument,       NULL, 'l' },
		{ "maintainer",  no_argument,       NULL, 'm' },
		{ "name",        no_argument,       NULL, 'n' },
		{ "or",          no_argument,       NULL, 'o' },
		{ "search_long", no_argument,       NULL, 's' },
		{ "inverse",     required_argument, NULL, 'v' },
		{ NULL,          no_argument,       NULL, 0 }
	};
	
	while (int ch = getopt_long(argc, argv, "Vhc:f:lmnosv:", longopts, NULL)) {
		if (-1 == ch) { break; }
		switch (ch) {
			case 'V': print_copyright(); return 0;
			case 'h': print_help(argv[0], index_filename); return 0;
			case 'c': categories.insert(string(optarg)); break;
			case 'f': index_filename = optarg; break;
			case 'l': flag_long = true; break;
			case 'm': flag_maintainer = true; break;
			case 'n': flag_name = true; break;
			case 'o': flag_or = true; break;
			case 's': flag_search_long = true; break;
			case 'v': if (not add_pattern(inverse_patterns, optarg)) { return 1; }; break;
			case '?':
			default: print_help(argv[0], index_filename); return 1;
		}
	}
	
	if (index_filename.empty()) {
		cerr << "Error: cannot determine the default path of the index file. Please specify" << endl;
		cerr << "the path to the index file manually using the -f option." << endl;
		return 1;
	}
	
	for (int i = optind; i < argc; ++i) {
		if (not add_pattern(patterns, argv[i])) { return 1; }
	}
	
	Index index(index_filename, !flag_name, flag_long | flag_search_long, flag_maintainer,
				not categories.empty());
	if (not index.readable()) {
		cerr << "Error reading index file \"" << index_filename << "\": "
			<< strerror(errno) << endl;
		return 1;
	}
	
	// finished with initalization
	
	// Now, read each line from the index file and maybe match it
	while (index.read_line()) {
		if (not index.parse_line()) { continue; }
		// Filter by categories, if the user specified it.
		if (not categories.empty()) {
			if (disjoint(categories, index.categories())) { continue; }
		}
		
		// Filter by inverse patterns, if the user specified it
		bool inv_match = false;
		for (const regex_t &r : inverse_patterns) {
			inv_match = regex_search(index.pkgname(), r)
				or regex_search(index.origin(), r)
				or regex_search(index.desc(), r)
				or (flag_maintainer and regex_search(index.maintainer(), r))
				or (flag_search_long and regex_search_file(index.descpath(), r));
			if (inv_match) { break; }
		}		
		if (inv_match) { continue; }
		
		// Filter by patterns
		bool match_one = false;
		bool match_all = true;
		for (const regex_t &r : patterns) {
			bool match = regex_search(index.pkgname(), r) 
					or regex_search(index.origin(), r)
					or regex_search(index.desc(), r)
					or (flag_maintainer and regex_search(index.maintainer(), r))
					or (flag_search_long and regex_search_file(index.descpath(), r));
			match_one |= match;
			match_all &= match;
			
			if (flag_or) {
				if (match_one) { break; }
			} else if (not match_all) {
				break;
			}
		}
		
		if (match_all or (flag_or and match_one)) {
			index.print_line(flag_name, flag_maintainer, flag_long);
		}
	}
	
	return 0;
}


void print_copyright() {
	cout << PROGNAME << ' ' << VERSION << endl;
	cout << COPYRIGHT << ' ' << AUTHOR << '@' << AUTHOR2 << endl;
}


void print_help(char* prog_name, const string &index_filename) {
	cout << "Usage: " << prog_name << R"( [options] PATTERN [PATTERN ...]

Lists ports whose description matches PATTERN. 

Options:
  -V, --version        Show program's version number and exit.
  -h, --help           Show this help message and exit.
  -c CATEGORY, --category=CATEGORY
                       Only search for ports in CATEGORY.
  -f FILE, --file=FILE Path to INDEX file. Default: ")" << index_filename << R"(".
  -l, --long           Display long description (pkg-descr file) for any match.
  -m, --maintainer     Search and display the maintainer.
  -n, --name           Print canonical name of a port, including its version.
  -o, --or             Search for ports that match any PATTERN.
  -s, --search_long    Search long descriptions (pkg-descr file). Slow.
  -v INVERSE_PATTERN, --inverse=INVERSE_PATTERN
                       Searches for ports that do not match a pattern.
)";
}


bool add_pattern(vector<regex_t> &patterns, const char *pattern) {
	constexpr int regex_flags = REG_EXTENDED | REG_ICASE | REG_NOSUB;
	regex_t regex;
	int retval = regcomp(&regex, pattern, regex_flags);
	if (retval) {
		char buffer[256];
		regerror(retval, &regex, buffer, sizeof(buffer));
		cerr << "Error in pattern: " << buffer << endl;
		return false;
	} else {
		patterns.push_back(regex);
		return true;
	}
}


bool disjoint(const set<string>& a, const set<string>& b) {
	set<string>::const_iterator it_a, it_b;
	for (it_a = a.begin(); it_a != a.end(); ++it_a) {
		for (it_b = b.begin(); it_b != b.end(); ++it_b) {
			if (*it_a == *it_b) { return false; }
		}
	}
	return true;
}


bool regex_search(const string &s, const regex_t &r) {
	return regexec(&r, s.c_str(), 0, NULL, 0) != REG_NOMATCH;
}


bool regex_search_file(const string &path, const regex_t &r) {
	string line;
	
	ifstream file(path);
	while (file.good()) {
		getline(file, line);
		if (regex_search(line, r)) { return true; }
	}
	return false;	
}


void set_default_index_filename(string &index_filename) {
#if defined(__FreeBSD__)
	// get kern.ostype sysctl to see whether we're running
	// on FreeBSD. If we're not, don't set a standard index
	// file path
	int mib[2] = { CTL_KERN, KERN_OSTYPE };
	char ostype[8];
	size_t len = sizeof(ostype);
	
	int retval = sysctl(mib, 2, ostype, &len, NULL, 0);
	if (retval != 0 && errno != ENOMEM) {
		perror("sysctl");
		return;
	}
	ostype[len - 1] = '\0';
	
	if (0 != strncmp("FreeBSD", ostype, len)) { return; }
	
	// get kern.osrelease sysctl to see whether we want
	// INDEX, INDEX-5 or INDEX-6 etc.
	mib[1] = KERN_OSRELEASE;
	char osrelease[32];
	len = sizeof(osrelease);
	retval = sysctl(mib, 2, osrelease, &len, NULL, 0);
	if (retval != 0 && errno != ENOMEM) {
		perror("sysctl");
		return;
	}
	osrelease[len - 1] = '\0';
	
	// get major version
	char* s = strchr(osrelease, '.');
	if (s == NULL) { return; }
	size_t major_version_str_len = (size_t)(s - osrelease) + 1;
	if (major_version_str_len < 1 or major_version_str_len > 3) { return; }
	char major_version_str[major_version_str_len];
	strlcpy(major_version_str, osrelease, major_version_str_len);
	
	index_filename = "/usr/ports/INDEX-";
	index_filename.append(major_version_str);
#else
	index_filename.clear();
#endif
}
